transform_dialect/examples/cuda/pad_via_pack_pipeline_spec.mlir

```diff
@@ -0,0 +1,218 @@
+// ```
+
+transform.sequence failures(propagate) {
+^bb1(%variant_op: !pdl.operation):
+  transform.print %variant_op : !pdl.operation
+
+  // Step 1. Convert pack/unpack to pad/extract_slice.
+  %pack = transform.structured.match ops{["tensor.pack"]} in %variant_op
+    : (!pdl.operation) -> !transform.op<"tensor.pack">
+  transform.structured.lower_pack %pack : (!transform.op<"tensor.pack">)
+    -> (!transform.op<"tensor.pad">, !transform.op<"tensor.expand_shape">,
+        !transform.op<"linalg.transpose">)
+  %unpack = transform.structured.match ops{["tensor.unpack"]} in %variant_op
+    : (!pdl.operation) -> !transform.op<"tensor.unpack">
+  transform.structured.lower_unpack %unpack : (!transform.op<"tensor.unpack">)
+    -> (!transform.op<"tensor.empty">,
+        !transform.op<"linalg.transpose">,
+        !transform.op<"tensor.collapse_shape">,
+        !transform.op<"tensor.extract_slice">)
+
+  transform.print %variant_op {name = "after conversion to pad"} : !pdl.operation
+
+  // Step 2. Tile and distribute.
+  %pad = transform.structured.match ops{["tensor.pad"]} in %variant_op
+    : (!pdl.operation) -> !pdl.operation
+
+  %forall_pad, %tiled_pad = transform.structured.tile_to_forall_op %pad
+    tile_sizes [8, 32] ( mapping = [#gpu.block<y>, #gpu.block<x>] )
+  transform.iree.populate_workgroup_count_region_using_num_threads_slice
+    %forall_pad : (!pdl.operation) -> ()
+
+  %if = transform.structured.match ops{["scf.if"]} in %forall_pad
+    : (!pdl.operation) -> !transform.any_op
+  transform.scf.take_assumed_branch %if take_else_branch
+    : (!transform.any_op) -> ()
+
+  transform.iree.apply_patterns %variant_op
+    {canonicalization, cse, licm, tiling_canonicalization} : (!pdl.operation) -> ()
+
+  transform.print %variant_op {name = "after tiling"} : !pdl.operation
+
+  // Step 3. Vectorize pad.
+  %pad_inside = transform.structured.match ops{["tensor.pad"]} in %forall_pad
+    : (!pdl.operation) -> !pdl.operation
+
+  transform.structured.masked_vectorize %pad_inside vector_sizes [8, 32]
+
+  transform.iree.apply_patterns %variant_op
+   {canonicalization, cse, licm, tiling_canonicalization} : (!pdl.operation) -> ()
+
+  transform.print %variant_op {name = "after masked vectorization"} : !pdl.operation
+
+  // Step 4. Rank-reduce and vectorize.
+  // ==================================
+  // Lower the masks to allow canonicalizations to kick in.
+  %func_v = transform.structured.match ops{["func.func"]} in %variant_op
+    : (!pdl.operation) -> !pdl.operation
+  %func_v_2 = transform.vector.lower_masked_transfers %func_v
+    : (!pdl.operation) -> !pdl.operation
+  transform.iree.apply_patterns %func_v_2 { rank_reducing_linalg, rank_reducing_vector }
+    : (!pdl.operation) -> ()
+  %func_v_3 = transform.structured.vectorize %func_v_2
+
+  transform.print %func_v_3 {name = "after vectorization"} : !pdl.operation
+
+  // Step 5. Bufferize and drop HAL descriptor from memref ops.
```

Comment:
Could we just drop this and everything below for now to land and iterate from a common state?

Reply:
done